src/Features/CSharp/Portable/CodeRefactorings/UseExplicitType/UseExplicitTypeCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,148 @@
+            var textSpan = context.Span;
+            var cancellationToken = context.CancellationToken;
+
+            if (!textSpan.IsEmpty)
+            {
+                return;
+            }
+
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+            var declaration = GetDeclaration(root, textSpan);
+            Debug.Assert(declaration.IsKind(SyntaxKind.VariableDeclaration, SyntaxKind.ForEachStatement, SyntaxKind.DeclarationExpression));
+
+            TypeSyntax declaredType = s_useExplicitTypeHelper.CanOfferFix(declaration, semanticModel, cancellationToken);
+            if (declaredType == null)
+            {
+                return;
+            }
+
+            State state = State.Generate(declaration, semanticModel, optionSet,
+                isVariableDeclarationContext: declaration.IsKind(SyntaxKind.VariableDeclaration), cancellationToken: cancellationToken);
+
+            // UseExplicitType analyzer/fixer already gives an action in this case
+            if (s_useExplicitTypeHelper.IsStylePreferred(semanticModel, optionSet, state, cancellationToken))
+            {
+                return;
+            }
+
+            Debug.Assert(state != null, "analyzing a declaration and state is null.");
+            if (!s_useExplicitTypeHelper.TryAnalyzeVariableDeclaration(declaredType, semanticModel, optionSet, cancellationToken))
+            {
+                return;
+            }
+
+            MakeRefactoring(context, document, declaredType, textSpan, cancellationToken);
+        }
+
+        private static SyntaxNode GetDeclaration(SyntaxNode root, TextSpan textSpan)
+        {
+            // Ex:
+            // var (x, y) = ...
+            // (var x, ...) = ...
+            // foreach (var (x, y) in ...) ...
+            var declarationExpression = root.FindToken(textSpan.Start).GetAncestor<DeclarationExpressionSyntax>();
+            if (declarationExpression != null)
+            {
+                return declarationExpression;
+            }
+
+            // Ex:
+            // var x = ...;
+            // var x;
+            // for (var x ... ) ...
+            // using (var x = ...) ...
+            var localDeclaration = root.FindToken(textSpan.Start).GetAncestor<VariableDeclarationSyntax>();
+            if (localDeclaration != null)
+            {
+                return localDeclaration;
+            }
+
+            // Ex:
+            // foreach (var x in ...) ...
+            var foreachStatement = root.FindToken(textSpan.Start).GetAncestor<ForEachStatementSyntax>();
```

Comment:
this seems super broad.  What makes it so that we're not offering this *way* too much just because we're deeply found underneath it?